File: csharp-server/GrpcServer/Services/UserServiceImpl.cs
using Grpc.Core;
using GrpcExample.UserService;
using System.Collections.Concurrent;

namespace GrpcServer.Services;

public class UserServiceImpl : UserService.UserServiceBase
{
    private readonly ILogger<UserServiceImpl> _logger;
    private readonly string _serverId;
    private readonly DateTime _startTime;
    private int _requestCount;
    
    // In-memory user store for demo purposes 
    private readonly ConcurrentDictionary<int, User> _users = new();
    private int _nextUserId = 1;

    public UserServiceImpl(ILogger<UserServiceImpl> logger)
    {
        _logger = logger;
        _serverId = $"CSHARP-{Environment.MachineName}-{Environment.ProcessId}-{Guid.NewGuid().ToString()[..8]}";
        _startTime = DateTime.UtcNow;
        
        // Seed some initial data
        SeedInitialData();
    }

    private void SeedInitialData()
    {
        var users = new[]
        {
            new User { Id = 1, Name = "Alice Johnson", Email = "alice@company.com", Department = "Engineering", CreatedTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds() },
            new User { Id = 2, Name = "Bob Smith", Email = "bob@company.com", Department = "Marketing", CreatedTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds() },
            new User { Id = 3, Name = "Carol Davis", Email = "carol@company.com", Department = "Engineering", CreatedTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds() },
            new User { Id = 4, Name = "David Wilson", Email = "david@company.com", Department = "Sales", CreatedTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds() },
            new User { Id = 5, Name = "Eva Brown", Email = "eva@company.com", Department = "Engineering", CreatedTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds() }
        };

        foreach (var user in users)
        {
            _users[user.Id] = user;
        }
        _nextUserId = 6;
    }

    public override Task<GetUserResponse> GetUser(GetUserRequest request, ServerCallContext context)
    {
        Interlocked.Increment(ref _requestCount);
        _logger.LogInformation("[{ServerId}] GetUser called for ID: {RequestedUserId}", _serverId, request.UserId);

        var response = new GetUserResponse { ServerId = _serverId };
        
        if (_users.TryGetValue(request.UserId, out var user))
        {
            response.User = user;
        }
        
        return Task.FromResult(response);
    }

    public override Task<CreateUserResponse> CreateUser(CreateUserRequest request, ServerCallContext context)
    {
        Interlocked.Increment(ref _requestCount);
        _logger.LogInformation("[{ServerId}] CreateUser called for: {RequestedName}", _serverId, request.Name);

        var user = new User
        {
            Id = Interlocked.Increment(ref _nextUserId),
            Name = request.Name,
            Email = request.Email,
            Department = request.Department,
            CreatedTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
        };

        _users[user.Id] = user;

        var response = new CreateUserResponse
        {
            User = user,
            ServerId = _serverId,
            Success = true,
            Message = "User created successfully"
        };

        return Task.FromResult(response);
    }

    public override async Task GetUsersByDepartment(GetUsersByDepartmentRequest request, IServerStreamWriter<User> responseStream, ServerCallContext context)
    {
        Interlocked.Increment(ref _requestCount);
        _logger.LogInformation("[{ServerId}] GetUsersByDepartment called for: {RequestedDepartment}", _serverId, request.Department);

        var users = _users.Values
            .Where(u => string.Equals(u.Department, request.Department, StringComparison.OrdinalIgnoreCase))
            .Take(request.Limit > 0 ? request.Limit : int.MaxValue);

        foreach (var user in users)
        {
            await responseStream.WriteAsync(user);
            await Task.Delay(100); // Simulate some processing time
        }
    }

    public override async Task<BulkCreateUsersResponse> BulkCreateUsers(IAsyncStreamReader<CreateUserRequest> requestStream, ServerCallContext context)
    {
        Interlocked.Increment(ref _requestCount);
        _logger.LogInformation("[{_serverId}] BulkCreateUsers called", _serverId);

        var createdUsers = new List<int>();
        var errors = new List<string>();
        var count = 0;

        await foreach (var request in requestStream.ReadAllAsync())
        {
            try
            {
                var user = new User
                {
                    Id = Interlocked.Increment(ref _nextUserId),
                    Name = request.Name,
                    Email = request.Email,
                    Department = request.Department,
                    CreatedTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
                };

                _users[user.Id] = user;
                createdUsers.Add(user.Id);
                count++;
            }
            catch (Exception ex)
            {
                errors.Add($"Failed to create user {request.Name}: {ex.Message}");
            }
        }

        var response = new BulkCreateUsersResponse
        {
            CreatedCount = count,
            ServerId = _serverId
        };
        
        response.UserIds.AddRange(createdUsers);
        response.Errors.AddRange(errors);

        return response;
    }

    public override async Task UserUpdatesStream(IAsyncStreamReader<UserUpdateRequest> requestStream, IServerStreamWriter<UserUpdateResponse> responseStream, ServerCallContext context)
    {
        Interlocked.Increment(ref _requestCount);
        _logger.LogInformation("[{ServerId}] UserUpdatesStream called", _serverId);

        await foreach (var request in requestStream.ReadAllAsync())
        {
            var response = new UserUpdateResponse { ServerId = _serverId };

            if (_users.TryGetValue(request.UserId, out var user))
            {
                // Update user based on the update type
                switch (request.UpdateTypeCase)
                {
                    case UserUpdateRequest.UpdateTypeOneofCase.NewName:
                        user.Name = request.NewName;
                        break;
                    case UserUpdateRequest.UpdateTypeOneofCase.NewEmail:
                        user.Email = request.NewEmail;
                        break;
                    case UserUpdateRequest.UpdateTypeOneofCase.NewDepartment:
                        user.Department = request.NewDepartment;
                        break;
                }

                response.UpdatedUser = user;
                response.Success = true;
                response.Message = "User updated successfully";
            }
            else
            {
                response.Success = false;
                response.Message = $"User with ID {request.UserId} not found";
            }

            await responseStream.WriteAsync(response);
        }
    }

    public override Task<ServerInfoResponse> GetServerInfo(ServerInfoRequest request, ServerCallContext context)
    {
        Interlocked.Increment(ref _requestCount);
        _logger.LogInformation("[{ServerId}] GetServerInfo called", _serverId);

        var uptime = (long)(DateTime.UtcNow - _startTime).TotalSeconds;
        
        var response = new ServerInfoResponse
        {
            ServerId = _serverId,
            ServerType = "C#",
            Version = "1.0.0",
            UptimeSeconds = uptime,
            TotalRequestsHandled = _requestCount
        };

        response.SupportedFeatures.AddRange([ 
            "Unary RPC", 
            "Server Streaming", 
            "Client Streaming", 
            "Bidirectional Streaming",
            "Health Checking",
            "Load Balancing"
        ]);

        return Task.FromResult(response);
    }
}

File: java-server/src/main/java/com/grpcexample/userservice/UserServiceImpl.java
package com.grpcexample.userservice;

import com.grpcexample.userservice.UserServiceGrpc.UserServiceImplBase;
import com.grpcexample.userservice.UserServiceProto.*;
import io.grpc.stub.StreamObserver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.List;
import java.util.ArrayList;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Java implementation of the UserService gRPC service
 * Demonstrates all gRPC communication patterns: unary, server streaming, client streaming, and bidirectional streaming
 */
public class UserServiceImpl extends UserServiceImplBase {
    private static final Logger logger = LoggerFactory.getLogger(UserServiceImpl.class);
    
    private final String serverId;
    private final long startTime;
    private final AtomicLong requestCount = new AtomicLong(0);
    private final ConcurrentHashMap<Integer, User> users = new ConcurrentHashMap<>();
    private final AtomicInteger nextUserId = new AtomicInteger(1000); // Start Java users at 1000 to distinguish from C#
    
    public UserServiceImpl(String serverId) {
        this.serverId = serverId;
        this.startTime = Instant.now().getEpochSecond();
        initializeTestData();
    }
    
    private void initializeTestData() {
        // Add some test users
        long timestamp = Instant.now().getEpochSecond();
        
        User user1 = User.newBuilder()
                .setId(1001)
                .setName("Alice Java")
                .setEmail("alice.java@company.com")
                .setDepartment("Engineering")
                .setCreatedTimestamp(timestamp)
                .build();
                
        User user2 = User.newBuilder()
                .setId(1002)
                .setName("Bob Java")
                .setEmail("bob.java@company.com")
                .setDepartment("Marketing")
                .setCreatedTimestamp(timestamp)
                .build();
                
        User user3 = User.newBuilder()
                .setId(1003)
                .setName("Carol Java")
                .setEmail("carol.java@company.com")
                .setDepartment("Engineering")
                .setCreatedTimestamp(timestamp)
                .build();
                
        User user4 = User.newBuilder()
                .setId(1004)
                .setName("David Java")
                .setEmail("david.java@company.com")
                .setDepartment("Sales")
                .setCreatedTimestamp(timestamp)
                .build();
                
        User user5 = User.newBuilder()
                .setId(1005)
                .setName("Eva Java")
                .setEmail("eva.java@company.com")
                .setDepartment("Engineering")
                .setCreatedTimestamp(timestamp)
                .build();
                
        users.put(user1.getId(), user1);
        users.put(user2.getId(), user2);
        users.put(user3.getId(), user3);
        users.put(user4.getId(), user4);
        users.put(user5.getId(), user5);
        
        nextUserId.set(1006);
        
        logger.info("Initialized test data with {} users", users.size());
    }
      
    @Override
    public void getUser(GetUserRequest request, StreamObserver<GetUserResponse> responseObserver) {
        requestCount.incrementAndGet();
        logger.info("[{}] GetUser called for ID: {}", serverId, request.getUserId());
        
        GetUserResponse.Builder responseBuilder = GetUserResponse.newBuilder()
                .setServerId(serverId);
                
        User user = users.get(request.getUserId());
        if (user != null) {
            responseBuilder.setUser(user);
        }
        
        responseObserver.onNext(responseBuilder.build());
        responseObserver.onCompleted();
    }
    
    @Override
    public void createUser(CreateUserRequest request, StreamObserver<CreateUserResponse> responseObserver) {
        requestCount.incrementAndGet();
        logger.info("[{}] CreateUser called for: {}", serverId, request.getName());
        
        try {
            User user = User.newBuilder()
                    .setId(nextUserId.getAndIncrement())
                    .setName(request.getName())
                    .setEmail(request.getEmail())
                    .setDepartment(request.getDepartment())
                    .setCreatedTimestamp(Instant.now().getEpochSecond())
                    .build();
                    
            users.put(user.getId(), user);
            
            CreateUserResponse response = CreateUserResponse.newBuilder()
                    .setUser(user)
                    .setServerId(serverId)
                    .setSuccess(true)
                    .setMessage("User created successfully")
                    .build();
                    
            responseObserver.onNext(response);
            responseObserver.onCompleted();
            
            logger.info("[{}] Created user with ID: {}", serverId, user.getId());
        } catch (Exception e) {
            logger.error("Error creating user", e);
            
            CreateUserResponse response = CreateUserResponse.newBuilder()
                    .setServerId(serverId)
                    .setSuccess(false)
                    .setMessage("Error creating user: " + e.getMessage())
                    .build();
                    
            responseObserver.onNext(response);
            responseObserver.onCompleted();
        }
    }
    
    @Override
    public void getUsersByDepartment(GetUsersByDepartmentRequest request, StreamObserver<User> responseObserver) {
        requestCount.incrementAndGet();
        logger.info("[{}] GetUsersByDepartment called for: {}", serverId, request.getDepartment());
        
        try {
            int limit = request.getLimit() > 0 ? request.getLimit() : Integer.MAX_VALUE;
            int count = 0;
            
            for (User user : users.values()) {
                if (count >= limit) break;
                
                if (user.getDepartment().equalsIgnoreCase(request.getDepartment())) {
                    responseObserver.onNext(user);
                    count++;
                    
                    // Add a small delay to demonstrate streaming
                    try {
                        Thread.sleep(100);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        break;
                    }
                }
            }
            
            responseObserver.onCompleted();
            logger.info("[{}] Completed streaming {} users for department: {}", serverId, count, request.getDepartment());
        } catch (Exception e) {
            logger.error("Error streaming users by department", e);
            responseObserver.onError(e);
        }
    }
    
    @Override
    public StreamObserver<CreateUserRequest> bulkCreateUsers(StreamObserver<BulkCreateUsersResponse> responseObserver) {
        requestCount.incrementAndGet();
        logger.info("[{}] BulkCreateUsers called", serverId);
        
        return new StreamObserver<CreateUserRequest>() {
            private final List<Integer> createdUserIds = new ArrayList<>();
            private final List<String> errors = new ArrayList<>();
            
            @Override
            public void onNext(CreateUserRequest request) {
                try {
                    User user = User.newBuilder()
                            .setId(nextUserId.getAndIncrement())
                            .setName(request.getName())
                            .setEmail(request.getEmail())
                            .setDepartment(request.getDepartment())
                            .setCreatedTimestamp(Instant.now().getEpochSecond())
                            .build();
                            
                    users.put(user.getId(), user);
                    createdUserIds.add(user.getId());
                    
                    logger.info("[{}] Created user via bulk stream: {} (ID: {})", serverId, user.getName(), user.getId());
                } catch (Exception e) {
                    String errorMsg = "Failed to create user " + request.getName() + ": " + e.getMessage();
                    errors.add(errorMsg);
                    logger.error(errorMsg, e);
                }
            }
            
            @Override
            public void onError(Throwable t) {
                logger.error("[{}] Error in bulkCreateUsers stream", serverId, t);
            }
            
            @Override
            public void onCompleted() {
                BulkCreateUsersResponse.Builder responseBuilder = BulkCreateUsersResponse.newBuilder()
                        .setCreatedCount(createdUserIds.size())
                        .setServerId(serverId);
                        
                responseBuilder.addAllUserIds(createdUserIds);
                responseBuilder.addAllErrors(errors);
                        
                responseObserver.onNext(responseBuilder.build());
                responseObserver.onCompleted();
                
                logger.info("[{}] Completed bulkCreateUsers - created {} users with {} errors", 
                           serverId, createdUserIds.size(), errors.size());
            }
        };
    }
    
    @Override
    public StreamObserver<UserUpdateRequest> userUpdatesStream(StreamObserver<UserUpdateResponse> responseObserver) {
        requestCount.incrementAndGet();
        logger.info("[{}] UserUpdatesStream bidirectional stream started", serverId);
        
        return new StreamObserver<UserUpdateRequest>() {
            @Override
            public void onNext(UserUpdateRequest request) {
                try {
                    logger.info("[{}] Received user update for ID: {}", serverId, request.getUserId());
                    
                    UserUpdateResponse.Builder responseBuilder = UserUpdateResponse.newBuilder()
                            .setServerId(serverId);
                    
                    User user = users.get(request.getUserId());
                    if (user != null) {
                        // Create updated user
                        User.Builder userBuilder = user.toBuilder();
                        
                        switch (request.getUpdateTypeCase()) {
                            case NEW_NAME:
                                userBuilder.setName(request.getNewName());
                                break;
                            case NEW_EMAIL:
                                userBuilder.setEmail(request.getNewEmail());
                                break;
                            case NEW_DEPARTMENT:
                                userBuilder.setDepartment(request.getNewDepartment());
                                break;
                            default:
                                responseBuilder.setSuccess(false)
                                              .setMessage("Unknown update type");
                                responseObserver.onNext(responseBuilder.build());
                                return;
                        }
                        
                        User updatedUser = userBuilder.build();
                        users.put(updatedUser.getId(), updatedUser);
                        
                        responseBuilder.setUpdatedUser(updatedUser)
                                      .setSuccess(true)
                                      .setMessage("User updated successfully");
                    } else {
                        responseBuilder.setSuccess(false)
                                      .setMessage("User with ID " + request.getUserId() + " not found");
                    }
                    
                    responseObserver.onNext(responseBuilder.build());
                } catch (Exception e) {
                    logger.error("[{}] Error processing user update", serverId, e);
                    
                    UserUpdateResponse errorResponse = UserUpdateResponse.newBuilder()
                            .setServerId(serverId)
                            .setSuccess(false)
                            .setMessage("Error processing update: " + e.getMessage())
                            .build();
                    responseObserver.onNext(errorResponse);
                }
            }
            
            @Override
            public void onError(Throwable t) {
                logger.error("[{}] Error in userUpdatesStream", serverId, t);
            }
            
            @Override
            public void onCompleted() {
                responseObserver.onCompleted();
                logger.info("[{}] UserUpdatesStream bidirectional stream completed", serverId);
            }
        };
    }
    
    @Override
    public void getServerInfo(ServerInfoRequest request, StreamObserver<ServerInfoResponse> responseObserver) {
        requestCount.incrementAndGet();
        logger.info("[{}] GetServerInfo called", serverId);
        
        long uptime = Instant.now().getEpochSecond() - startTime;
          ServerInfoResponse.Builder responseBuilder = ServerInfoResponse.newBuilder()
                .setServerId(serverId)
                .setServerType("Java")
                .setVersion("1.0.0")
                .setUptimeSeconds(uptime)
                .setTotalRequestsHandled((int) requestCount.get());
                
        responseBuilder.addSupportedFeatures("Unary RPC");
        responseBuilder.addSupportedFeatures("Server Streaming");
        responseBuilder.addSupportedFeatures("Client Streaming");
        responseBuilder.addSupportedFeatures("Bidirectional Streaming");
        responseBuilder.addSupportedFeatures("Health Checking");
        responseBuilder.addSupportedFeatures("Load Balancing");
                
        responseObserver.onNext(responseBuilder.build());
        responseObserver.onCompleted();
    }
}
